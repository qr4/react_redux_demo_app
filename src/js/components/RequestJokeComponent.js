// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Actions } from '../state/Actions';
import type { Dispatch } from 'redux';

type RequestJokeComponentState = { input: string };

export class DisconnectedRequestJokeComponent extends Component<
    { requestJoke: number => void },
    RequestJokeComponentState
> {
    state = { input: '' };

    _onInputChange = (e: SyntheticInputEvent<*>) => {
        this.setState({
            input: e.target.value,
        });
    };

    _onSubmit = (e: SyntheticEvent<*>) => {
        e.preventDefault();
        const parsedInt = parseInt(this.state.input, 10);
        if (!isNaN(parsedInt)) {
            this.props.requestJoke(parsedInt);
            this.setState({ input: '' });
        }
    };

    render() {
        return (
            <form onSubmit={this._onSubmit}>
                <div className="form-group">
                    <label className="control-label">Request more Jokes!</label>
                    <input
                        type="number"
                        className="form-control"
                        value={this.state.input}
                        onChange={this._onInputChange}
                        placeholder="Enter id..."
                    />
                </div>

                <button className="btn btn-primary">Request joke with this id</button>
            </form>
        );
    }
}

const mapDispatchToProps = (dispatch: Dispatch) => ({ requestJoke: num => dispatch(Actions.requestJoke(num)) });

export const RequestJokeComponent = connect(null, mapDispatchToProps)(DisconnectedRequestJokeComponent);
