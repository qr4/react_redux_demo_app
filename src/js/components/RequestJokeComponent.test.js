// @flow

import React from 'react';
import { shallow } from 'enzyme';

import { DisconnectedRequestJokeComponent } from './RequestJokeComponent';

describe('RequestJokeComponent', () => {
    it('[snapshot] it renders correctly', () => {
        const app = shallow(<DisconnectedRequestJokeComponent requestJoke={() => {}} />);
        expect(app).toMatchSnapshot();
    });

    it('calls handler with correct value', () => {
        const spy = jest.fn();
        const preventDefaultSpy = jest.fn();
        const form = shallow(<DisconnectedRequestJokeComponent requestJoke={spy} />);

        const formControl = form.find('.form-control');

        // simulate the input change
        formControl.simulate('change', { target: { value: '123' } });

        form.simulate('submit', { preventDefault: preventDefaultSpy });

        expect(preventDefaultSpy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(123);
    });

    it('does not call handler if input was non-numeric', () => {
        const spy = jest.fn();
        const form = shallow(<DisconnectedRequestJokeComponent requestJoke={spy} />);

        const formControl = form.find('.form-control');

        // simulate the input change
        formControl.simulate('change', { target: { value: 'NOT_A_NUMBER' } });

        form.simulate('submit', { preventDefault: () => {} });

        expect(spy).not.toHaveBeenCalled();
    });
});
