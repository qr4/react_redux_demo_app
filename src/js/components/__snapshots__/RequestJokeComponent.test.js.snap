// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestJokeComponent [snapshot] it renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DisconnectedRequestJokeComponent
    requestJoke={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="form-group"
>
          <label
                    className="control-label"
          >
                    Request more Jokes!
          </label>
          <input
                    className="form-control"
                    onChange={[Function]}
                    placeholder="Enter id..."
                    type="number"
                    value=""
          />
</div>,
        <button
          className="btn btn-primary"
>
          Request joke with this id
</button>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              className="control-label"
>
              Request more Jokes!
</label>,
            <input
              className="form-control"
              onChange={[Function]}
              placeholder="Enter id..."
              type="number"
              value=""
/>,
          ],
          "className": "form-group",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Request more Jokes!",
              "className": "control-label",
            },
            "ref": null,
            "rendered": "Request more Jokes!",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control",
              "onChange": [Function],
              "placeholder": "Enter id...",
              "type": "number",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Request joke with this id",
          "className": "btn btn-primary",
        },
        "ref": null,
        "rendered": "Request joke with this id",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="form-group"
>
            <label
                        className="control-label"
            >
                        Request more Jokes!
            </label>
            <input
                        className="form-control"
                        onChange={[Function]}
                        placeholder="Enter id..."
                        type="number"
                        value=""
            />
</div>,
          <button
            className="btn btn-primary"
>
            Request joke with this id
</button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                className="control-label"
>
                Request more Jokes!
</label>,
              <input
                className="form-control"
                onChange={[Function]}
                placeholder="Enter id..."
                type="number"
                value=""
/>,
            ],
            "className": "form-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Request more Jokes!",
                "className": "control-label",
              },
              "ref": null,
              "rendered": "Request more Jokes!",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "onChange": [Function],
                "placeholder": "Enter id...",
                "type": "number",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Request joke with this id",
            "className": "btn btn-primary",
          },
          "ref": null,
          "rendered": "Request joke with this id",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
